docker/docker_image_dest.go

```diff
@@ -131,15 +132,49 @@ func (d *dockerImageDestination) HasThreadSafePutBlob() bool {
 // to any other readers for download using the supplied digest.
 // If stream.Read() at any time, ESPECIALLY at end of input, returns an error, PutBlob MUST 1) fail, and 2) delete any data stored so far.
 func (d *dockerImageDestination) PutBlob(ctx context.Context, stream io.Reader, inputInfo types.BlobInfo, cache types.BlobInfoCache, isConfig bool) (types.BlobInfo, error) {
+	digester, stream := putblobdigest.DigestIfCanonicalUnknown(stream, inputInfo)
+	sizeCounter := &sizeCounter{}
+	stream = io.TeeReader(stream, sizeCounter)
```

Comment:
The way these values are applied/computed in different locations is non-obvious and doesn‘t look necessary to me, e.g. the way `digester` _exists_ during the upload but might actually only be computed on the digest pre-computation path.

I’d prefer for these to be initialized where they were, and for the new code to have its own digester (which can just be a `digest.Canonical.Digester` without using a redundant `DigestIfCanonicalUnknown` condition); we also don’t need two locations of `sizeCounter` at all, AFAICS.